src/puppetlabs/dujour/version_check.clj

```diff
@@ -65,28 +71,72 @@
                          (dissoc :product-name)
+                         (assoc "product" artifact-id)
+                         (assoc "group" group-id)
                          (merge version-data)
-                         ring-codec/form-encode)
-        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
-        {:keys [status body] :as resp} (client/get url
+                         json/generate-string)
+        url update-server-url
+        {:keys [status body] :as resp} (client/post url
                                                    {:headers {"Accept" "application/json"}
+                                                    :body request-body
                                                     :as :text})]
-    (if (= status 200)
+    {:status status :body body :resp resp}))
+
+(schema/defn ^:always-validate update-info-get :- (schema/maybe RequestResult)
+  "Make a GET request to the puppetlabs server to determine the latest available
+  version."
+  [request-values :- RequestValues
+   update-server-url :- schema/Str]
+  (let [product-name (:product-name request-values)
+        {:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (str (or (:version request-values) (version group-id artifact-id "")))
+        version-data {:version current-version}
+        query-params (-> request-values
+                         (dissoc :product-name)
+                         (assoc :product artifact-id)
+                         (assoc :group group-id)
+                         (merge version-data)
+                         (walk/stringify-keys))
+        {:keys [status body] :as resp} (client/get update-server-url
+                                                   {:headers {"Accept" "application/json"}
+                                                    :query-params query-params
+                                                    :as :text})]
+    {:status status :body body :resp resp}))
+
+(schema/defn ^:always-validate update-info :- (schema/maybe UpdateInfo)
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Attempts a POST request before falling back to a GET request.
+  Returns the JSON object received from the server, which is expected to be
+  a map containing keys `:version`, `:newer`, and `:link`. Returns `nil` if
+  the request does not succeed for some reason."
+  [request-values :- RequestValues
+   update-server-url :- schema/Str]
+  (let [{:keys [status body resp]} (update-info-post request-values update-server-url)]
+    (cond
+      (= status 200)
       (json/parse-string body true)
+
+      ;; POST is a 404, try GET instead.
+      (= status 404)
+      (let [{:keys [status body resp]} (update-info-get request-values update-server-url)]
+        (if (= status 200)
+          (json/parse-string body true)
+          (sling/throw+ {:type ::update-request-failed
+                         :message resp})))
```

Comment:
I think it's probably okay to print the stack trace...the problem is the GET/POST requests in update-info throw the exception and it isn't caught any where so we never get to this point in the library. Instead there are a bunch of connection refused exceptions in the log. :(